src/core/lib/channel/metrics.h

```diff
@@ -87,37 +91,238 @@ class GlobalInstrumentsRegistry {
   using GlobalCallbackHandle = absl::variant<GlobalCallbackInt64GaugeHandle,
                                              GlobalCallbackDoubleGaugeHandle>;
 
+  template <typename ValueType, typename InstrumentType, std::size_t M,
+            std::size_t N>
+  struct MagicHandle : public GlobalInstrumentHandle {
```

Comment:
Not sure how this will work, the size information needs to be attached on the handle to allow compile-time check if the label value sizes are different from the label key sizes. There is another limitation that we cannot templatize the StatsPlugin interface since C++ does not support virtual template member function.

Reply:
Good point about needing the size information in the handle.  I think what I sketched out above can be changed to do that:

```
class GlobalInstrumentsRegistry {
 public:
  // ...

  template <ValueType V, InstrumentType I, size_t N, size_t M>
  struct TypedGlobalInstrumentHandle : public GlobalInstrumentHandle {
    static constexpr ValueType value_type = V;
    static constexpr InstrumentType instrument_type = I;
    static constexpr size_t num_labels = N;
    static constexpr size_t num_optional_labels = M;
  };

  template <ValueType V, InstrumentType I, std::size_t M, std::size_t N>
  class RegistrationBuilder {
   public:
    // ...

    TypedGlobalInstrumentHandle<V, I, N, M> Build() {
      TypedGlobalInstrumentHandle<V, I, N, M> handle;
      handle.index = GlobalInstrumentsRegistry::RegisterInstrument(
          V, I, description_, unit_, enable_by_default_, label_keys_,
          optional_label_keys_);
      return handle;
    }

    // ...
  };

  // Creates instrument in the GlobalInstrumentsRegistry.
  static RegistrationBuilder<ValueType::kUInt64, InstrumentType::kCounter, 0, 0>
  RegisterUInt64Counter(
      absl::string_view name, absl::string_view description,
      absl::string_view unit, bool enable_by_default) {
    return RegistrationBuilder<ValueType::kUInt64, InstrumentType::kCounter, 0, 0>(
        name, description, unit, enable_by_default);
  }

  // ...
};

class StatsPlugin {
 public:
  // ...

  virtual void AddCounter(
      GlobalInstrumentsRegistry::GlobalInstrumentHandle handle,
      uint64_t value, absl::Span<const absl::string_view> label_values,
      absl::Span<const absl::string_view> optional_label_values) = 0;

  // ...
};

class StatsPluginRegistry {
 public:
   // ...

  class StatsPluginGroup {
   public:
    // ...

    template <typename HandleType, typename NativeValueType,
              typename = typename std::enable_if<
                  std::is_same<HandleType::instrument_type, InstrumentType::kCounter>::value>::type>
    void AddCounter(HandleType handle, NativeValueType value,
                    std::array<absl::string_view, HandleType::num_labels> label_values,
                    std::array<absl::string_view, HandleType::num_optional_labels> optional_values) {
      for (auto& state : plugins_state_) {
        state.plugin->AddCounter(handle, value, label_values,
                                 optional_values);
      }
    }

    template <typename HandleType, typename NativeValueType,
              typename = typename std::enable_if<
                  std::is_same<HandleType::instrument_type, InstrumentType::kHistogram>::value>::type>
    void RecordHistogram(HandleType handle, NativeValueType value,
                    std::array<absl::string_view, HandleType::num_labels> label_values,
                    std::array<absl::string_view, HandleType::num_optional_labels> optional_values) {
      for (auto& state : plugins_state_) {
        state.plugin->RecordHistogram(handle, value, label_values,
                                 optional_values);
      }
    }

    // ...
  };

  // ...
};
```

Note that this means we're removing the convenience aliases (`GlobalUInt64CounterHandle` and friends) and changing the `StatsPlugin` methods to use the base `GlobalInstrumentHandle` type.  But there should be no need for a virtual templated function there.  (Note that gRPC components that define metrics don't directly use the `StatsPlugin` API; they go through `StatsPluginGroup`.  So as long as we have the right compile-time checking in `StatsPluginGroup`, I think we're fine.)

Reply:
Great suggestion! This indeed simplifies things a bit and good suggestion on using the ValueType and InstrumentType enum class as non-type template parameters. PTAL.